pgxn/neon/walproposer.c

```diff
@@ -1393,8 +1393,22 @@ WalProposerRecovery(int donor, TimeLineID timeline, XLogRecPtr startpos, XLogRec
 	char	   *err;
 	WalReceiverConn *wrconn;
 	WalRcvStreamOptions options;
+	char conninfo[MAXCONNINFO];
 
-	wrconn = walrcv_connect(safekeeper[donor].conninfo, false, "wal_proposer_recovery", &err);
+	if (!neon_auth_token)
+	{
+		memcpy(conninfo, safekeeper[donor].conninfo, MAXCONNINFO);
+	}
+	else
+	{
+		int written = 0;
+
+		written = snprintf((char *) conninfo, MAXCONNINFO, "password=%s %s", neon_auth_token, safekeeper[donor].conninfo);
```

Comment:
This is a hack and won't work if password have spaces. Anyone knows a better way to append password to connection string?

Reply:
I think you can %-encode special characters, e.g. spaces can become `%20`

https://datatracker.ietf.org/doc/html/rfc3986#section-2.1

Reply:
Postgres docs: https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING

Reply:
There is PQconninfoParse with which you can parse connection string to get all options separately. However then PQconnectStartParams must be used, which is untrivial with libpqwalreceiver as it is hidden inside. Better check how our tokens look like, but likely just quoting with '' would be enough (or even nothing).

Reply:
It all depends on the `neon_auth_token` format. If it will have only alphanumeric characters, I think it's fine as is. Otherwise, we either need to use some escaping function for this case specifically (I haven't found this function yet), or use `PQconninfoParse` + append password + `PQ_construct_connstr_from_kv (not found)`, but it also needs a function which I haven't found yet.

Reply:
On a quick glance JWT tokens are base64 encoded, so quotes should be enough.